### DISCUSSION TITLE
My go app that uses embed with wildcard fails to build

### DISCUSSION THREAD

**Discussion by daisukixci** [wailsapp/wails#2849] [Q&A]
Hi,

I started using Wails for an application, but I have an issue regarding embedding files (others than the static for the webapp) when there is a wildcard in the embed directive.

My tree looks like
```
❯ tree  backend/check
backend/check
├── checks
│   └── dummy
│       └── check.json
│       └── check.sh
├── fs.go
...
```

fs.go contains 
```go
package check

import (
	"embed"
...
)

//go:embed checks/*/check.json
var checksF embed.FS

//go:embed checks/*
var dataF embed.FS
...
```

When trying to run `wails dev`, I get 
```
❯ wails dev
Wails CLI v2.5.1

Executing: go mod tidy
  • Generating bindings:

  ERROR
          backend/check/fs.go:15:12: pattern checks/*: cannot embed directory checks/*: invalid name *

          exit status 1
 ♥   If Wails is useful to you or your company, please consider sponsoring the project:
https://github.com/sponsors/leaanthony
```
And I see that `backend/check/checks/*/` is created and contains:
```
tree backend/check/checks/\*
backend/check/checks/*
└── check.json
    └── gitkeep

1 directory, 1 file
```

Am I doing something wrong ? Could it be a bug ? I feel like it is related to the wildcard usage as when removed it does work

Thanks for time and help

**Comment by BillBuilt** [comment]
I encountered this problem awhile ago too. My Wails app had sat for several months and when I came back to it after updating all of my dependencies, including Wails, the wildcard stopped working. I was able to get around it but it is certainly undesired.

**Reply by daisukixci** [reply]
How did you get around?

**Reply by BillBuilt** [reply]
My case was simple. I have a folder with .sql migration files and  I only wanted to embed *.sql files from that folder (just to play it safe)
`//go:embed database/migrations/*.sql`

But when this popped up I had to just remove the wildcard and embed the entire directory:
`//go:embed database/migrations`

Considering I didn't have any other types of files in there it worked fine, but if I were to accidentally somehow get a non .sql file in there, it would fail horribly.

However this does eliminate the possibility of a README or any other type of non .sql file from ever being able to be placed in there.

**Reply by daisukixci** [reply]
In the meantime I did something similar

**Reply by leaanthony** [reply]
This is unrelated to Wails as it's purely a Go directive. Perhaps they stopped supporting wildcards?

**Reply by daisukixci** [reply]
Building without Wails works though

**Reply by BillBuilt** [reply]
We use wildcard embeds at my employment: `//go:embed mappings/*.json` and it works fine with Go 1.20.7 (just Go, no Wails).

Plenty of references about using wildcards and other patterns in the Go docs: https://pkg.go.dev/embed

Many of their examples show using wildcards:

```golang
package main

import (
	"embed"
	"log"
	"net/http"
)

//go:embed internal/embedtest/testdata/*.txt
var content embed.FS

func main() {
	mutex := http.NewServeMux()
	mutex.Handle("/", http.FileServer(http.FS(content)))
	err := http.ListenAndServe(":8080", mutex)
	if err != nil {
		log.Fatal(err)
	}
}
```

**Reply by leaanthony** [reply]
> Building without Wails works though

This is literally a Go only directive and has nothing to do with Wails.

**Reply by stffabi** [reply]
@leaanthony I suspect the problem is due to our code that makes sure the embed directories exist. So in this case `//go:embed checks/*` we create a directory structure like this `checks/*/.gitkeep`. And embedding a folder or file with name `*` is invalid. That's why the go build will then fail with: `invalid name *`

**Reply by leaanthony** [reply]
Ah that makes a ton more sense 😕
